### DISCUSSION TITLE
How to inject an APK as system app WITHOUT using external utils like Magisk?

### DISCUSSION THREAD

**Discussion by TriMoon** [waydroid/waydroid#1001] [Q&A] [answered]
Where do we place apk files we want to be auto installed as system apps?

I have an apk from some other Androidx86 version for keyboard mapping, i think was from DarkMatter, that i want to try in waydroid that needs to be installed as system app.
- CZGameHelper: [GameHelper.apk.gz](https://github.com/waydroid/waydroid/files/12029798/GameHelper.apk.gz)
_(You need to `gunzip` this file first, it was `gzip`'ed to be able to post here...)_

---
## NOT working:
- apk at `~/.local/share/waydroid/data/system/priv-app/GameHelper/GameHelper.apk`

**Comment by ChrysoliteAzalea** [comment]
You can "systemlessly" inject it [using Magisk](https://github.com/nitanmarcel/waydroid-magisk/) and then creating a simple Magisk module that adds your app as a system app (that is safer than editing the image since it doesn't actually modify it and can be rolled back easily).

**Comment by TriMoon** [comment]
My question was not "how to install using extra utils" but thanks for the effort... :+1:

**Reply by ChrysoliteAzalea** [reply]
Sorry. The answer is, to create a directory under ```/system/priv-app/``` and put an .apk there. However, I strongly recommend doing it through Magisk (or overlay) in order to avoid patching the image.

**Reply by ChrysoliteAzalea** [reply]
The module should contain all files and directories, that is to be overlay-mounted to the system. (for example, you can try creating a zip archive with a ```/system/priv-app/MyApp/App.apk``` path, and installing it as a Magisk module)

**Reply by TriMoon** [reply]
> @ChrysoliteAzalea The answer is, to create a directory under `/system/priv-app/` and put an .apk there.

Thanks, i will try to put the apk in the overlay at `~/.local/share/waydroid/data/system/priv-app/MyAppName/`, **without Magisk**, and see in due time when i try it out :+1: 

_Again the question was about a way WITHOUT using external utils..._

**Reply by ChrysoliteAzalea** [reply]
Alternatively, you can enlarge an image (for example, ```tune2fs```), mount it and add the file to ```/system/priv-app/MyApp```.

**Reply by TriMoon** [reply]
You have a habbit of forcing people to sound irritated, but...
Which part of "**WITHOUT using external utils...**" don't you understand, maybe i can reword it? :woman_shrugging:

**Reply by ChrysoliteAzalea** [reply]
The problem is, that ```system.img``` image usually doesn't have enough free space for extra data, so it has to be enlarged. You can check how much free space it has after mounting it (but enlarging and another tuning operations should be performing while the image is **NOT** mounted).

**Reply by TriMoon** [reply]
space in `system.img` should be of no concern when using the overlay.
Anyhow i will try someday...

**Reply by TriMoon** [reply]
I finally found some time to test this but it isn't working...
I placed the apk at `~/.local/share/waydroid/data/system/priv-app/GameHelper/GameHelper.apk`

NOT working, the app is not to be found inside waydroid...

**Reply by ChrysoliteAzalea** [reply]
Well, that's placing the app in ```/data/system/priv-app```, not ```/system/priv-app/```. The system image and overlays are located in ```/var/lib/waydroid/```. The overlay directory you mean is likely ```/var/lib/waydroid/overlay/system/priv-app```.

**Reply by TriMoon** [reply]
ok ill try that tomorrow then

**Reply by TriMoon** [reply, marked as answer]
Solved in #1109, unfortunately non-working apk :woman_shrugging:

**Comment by Isabellae4567** [comment]
If you want to install an APK as a system app without using external utilities, you can place the APK manually in /system/app/ or /system/priv-app/ depending on whether it requires elevated privileges. After placing it, make sure the permissions are correct so the system recognizes it as a system app. However, this method can be risky because directly modifying system files may cause instability or boot issues. A safer and more flexible approach is to use [magisk](https://magiskapp.pro/), which allows systemless modifications on Android devices running 6.0 and above. With Magisk, you can safely manage system apps, install modules, and customize your Android device without altering core system files, reducing the chance of errors while still achieving the same results.

**Reply by TriMoon** [reply]
:point_right: https://github.com/waydroid/waydroid/discussions/1001#discussioncomment-6929255
